server/chat-plugins/abuse-monitor.ts

```diff
@@ -0,0 +1,727 @@
+	"TOXICITY": {},
+	"IDENTITY_ATTACK": {},
+	"INSULT": {},
+	"PROFANITY": {},
+	"THREAT": {},
+};
+
+export const cache: {
+	[roomid: string]: {[userid: string]: number} & {
+		staffNotified?: boolean,
+		claimed?: ID,
+	},
+} = global.Chat?.oldPlugins['abuse-monitor']?.cache || {};
+
+for (const k in cache) {
+	const entry = cache[k] as any;
+	if (entry.flags) delete entry.flags;
+	if (entry.users) {
+		cache[k] = {...entry.users, staffNotified: entry.notified};
+	}
+}
+
+const defaults: FilterSettings = {
+	threshold: 4,
+	minScore: 0.65,
+	specials: {
+		THREAT: {0.96: 'MAXIMUM'},
+		IDENTITY_ATTACK: {0.8: 2},
+		SEVERE_TOXICITY: {0.8: 2},
+	},
+};
+
+export const settings: FilterSettings = (() => {
+	try {
+		// accounting for data changes -
+		// make sure we do have the default data in case it's not in the stored data
+		return {...defaults, ...JSON.parse(FS('config/chat-plugins/nf.json').readSync())};
+	} catch (e: any) {
+		if (e.code !== "ENOENT") throw e;
+		return defaults;
+	}
+})();
+
+interface FilterSettings {
+	disabled?: boolean;
+	threshold: number;
+	minScore: number;
+	specials: {[k: string]: {[k: number]: number | "MAXIMUM"}};
+}
+
+export interface PerspectiveRequest {
+	languages: string[];
+	requestedAttributes: AnyObject;
+	comment: {text: string};
+}
+
+export interface PMResult {
+	score: number;
+	flags: string[];
+}
+
+function time() {
+	return Math.floor(Date.now() / 1000);
+}
+
+export class Limiter {
+	second = time();
+	countThisSecond = 0;
+	private counts = [] as number[];
+	private period: number;
+	max: number;
+	constructor(max: number, period: number) {
+		this.max = max;
+		this.period = period;
+	}
+	shouldRequest() {
+		const now = time();
+		if (this.second !== now) {
+			const diff = now - this.second;
+			if (diff <= this.period) {
+				for (let i = 0; i < (diff - 1); i++) {
+					this.counts.push(0);
+				}
+			} else {
+				this.counts = Array(this.period).fill(0);
+			}
+			this.counts.push(this.countThisSecond);
+			this.trimCounts();
+			this.second = now;
+			this.countThisSecond = 0;
+		}
+		const count = this.countThisSecond + 1;
+		const counts = this.getCounts();
+		const total = [...counts, count].reduce((a, b) => a + b);
+		const avg = total / (counts.length + 1);
+		// if it'd go over, queue instead
+		if (avg > this.max) {
+			return false;
+		}
+		// if it'd succeed, now we increment the count since
+		// we're going to allow it to happen
+		this.countThisSecond++;
+		return true;
+	}
+	/* private enqueue(depth = 0) {
+
+        return new Promise<boolean>(resolve => {
+            setTimeout(() => resolve(this.shouldRequest(depth + 1)), 1000);
+        });
+    }*/
+	getCounts() {
+		this.trimCounts();
+		return this.counts;
+	}
+	private trimCounts() {
+		if (this.counts.length > this.period) {
+			// oldest first
+			while (this.counts.length > this.period) this.counts.shift();
+		}
+	}
+}
+
+function isCommon(message: string) {
+	message = message.replace(/\?!\., ;:/ig, '');
```

Comment:
```suggestion
	message = message.toLowerCase().replace(/\?!\., ;:/g, '');
```

Reply:
Don't you want `/i`? What if someone types `WP`?